### PR TITLE
[IO3]Transformations on a driver level + visitor framework issues resolved

### PR DESCRIPTION
**Pull request by snapeeek**
Adam Wojtczak 215922
Maciej Woźniak 215925

### CODE REVIEW

**Comment by snapeeek** [comment]
Ostatnie kilka commitów dotyczy taska "Visitor framework in commands hierarchy"

File: src/main/java/edu/kis/powp/jobs2d/drivers/transformations/DriverLevelTransformation.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package edu.kis.powp.jobs2d.drivers.transformations;
+
+public interface DriverLevelTransformation {
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/drivers/transformations/DriverLevelTransformation.java]: There is nothing concerning drivers in this interface. CoordsTransformation or XYTransformation would be better.

File: src/main/java/edu/kis/powp/jobs2d/drivers/ScaledLineDriver.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Line adapter - Job2dDriver with DrawPanelController object.
+ */
+public class ScaledLineDriver implements Job2dDriver {
+	private ILine line;
+	private int startX = 0, startY = 0;
+	private String name;
+	private DriverLevelTransformation transformation;
+
+	private DrawPanelController drawController;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/drivers/ScaledLineDriver.java]: You should not reimplement a drawing driver again. Use the decorator pattern to re-use already existing drivers!

File: src/main/java/edu/kis/powp/jobs2d/drivers/ScaledLineDriver.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+import edu.kis.legacy.drawer.panel.DrawPanelController;
+import edu.kis.legacy.drawer.shape.ILine;
+import edu.kis.powp.jobs2d.Job2dDriver;
+import edu.kis.powp.jobs2d.drivers.transformations.DriverLevelTransformation;
+
+/**
+ * Line adapter - Job2dDriver with DrawPanelController object.
+ */
+public class ScaledLineDriver implements Job2dDriver {
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/drivers/ScaledLineDriver.java]: It is not only about Scaling and it does not need to be about lines only. Your goal should be, e.g., Driver2dTransformationsDecorator.

File: src/test/java/edu/kis/powp/jobs2d/events/SelectDriverTransformationRotating.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+{
+
+	@Override
+	public void actionPerformed(ActionEvent e)
+	{
+		DrawPanelController drawerController = DrawerFeature.getDrawerController();
+		Job2dDriver driver = null;
+		if (DriverFeature.getDriverManager().getCurrentDriver().getClass() == LineDriverAdapter.class)
+		{
+			LineDriverAdapter temp = (LineDriverAdapter)DriverFeature.getDriverManager().getCurrentDriver();
```
Comment (kowallus) [inline src/test/java/edu/kis/powp/jobs2d/events/SelectDriverTransformationRotating.java]: Incorrect and not object-oriented. What if the current driver is not a line?!

File: src/test/java/edu/kis/powp/jobs2d/events/SelectDriverTransformationScaling.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	{
+		DrawPanelController drawerController = DrawerFeature.getDrawerController();
+		Job2dDriver driver = null;
+		if (DriverFeature.getDriverManager().getCurrentDriver().getClass() == LineDriverAdapter.class)
+		{
+			LineDriverAdapter temp = (LineDriverAdapter)DriverFeature.getDriverManager().getCurrentDriver();
+			driver = new ScaledLineDriver(drawerController, temp.getLine(), "inherited", new DriverLevelScaleTransformation(1.5, 1));
+		}
+		DriverFeature.getDriverManager().setCurrentDriver(driver);
+	}
```
Comment (kowallus) [inline src/test/java/edu/kis/powp/jobs2d/events/SelectDriverTransformationScaling.java]: major code duplication between listeners

**Review by kowallus — CHANGES_REQUESTED** [review]
Try to fix the following issues.

**Comment by kowallus** [comment]
please close, fix and reopen

**Comment by kowallus** [comment]
> Ostatnie kilka commitów dotyczy taska "Visitor framework in commands hierarchy"

OK, ale takie poprawki lepiej robić w osobnych gałęziach i wystawiać jako osobny PR.

**Pull request merged**